Add special cases

verhovsky opened the pull request:
I created a script that extracts all the elements with an `id=` attribute in all the `.html` files in a directory. Then I ran it on the [archives of the documentation](https://www.python.org/ftp/python/doc/) of all the Python 2 minor versions and 3.8. Then I looked at which links are missing from 3.8 and manually figured out where they should point to, or if they should explicitly point to nothing with `null`. I had to change the redirect code: it checks if the link is a special case, then it checks if the link is in a file that is a special case and finally it falls back to keeping the part after `docs.python.org/2.x/` the same and just changing the version number. 

It would be good if users were able to at flag certain redirects as bad and suggest edits and post comments/justifications for a given redirect like the hash of the commit that changed a redirect (using `git bisect`) from the extension popup. It would also be good if the special cases could come with an optional text field for explaining why something was redirected where it was, what to use instead, etc. I had to figure a bunch of that out as I was working on this but I didn't add that knowledge to this extension.

I did what I could for all the 2.x versions, but there's tons of links (like 10's of thousands) in the older Python 2 versions (and some in 2.6 and 2.7) that can't be redirected without knowing which minor version you're redirecting from, because they're just generated numbers that change between minor versions in the documentation for example

* https://docs.python.org/2.2/lib/node216.html
* https://docs.python.org/2.5/lib/node216.html

and links like https://docs.python.org/2.5/lib/testresult-objects.html#l2h-5043 (and even `node<whatever>.html#l2h-<whatever>` links) which also probably change depending on the minor version, but I didn't check.

m4tx commented on background.js:
```diff
@@ -1,8 +1,2714 @@
     const SPECIAL_CASES = {
-        "/library/sets.html" : "/library/stdtypes.html#set",
-        "/library/stringio.html": "/library/io.html#io.StringIO"
+        // TODO: these commented links might break when 2to3 is removed
```
Maybe now it would make sense to move all of those `SPECIAL_CASES` to a separate file? `background.js` has become very large because of that.

verhovsky commented:
I got the redirection code into a state I'm happy with, so feel free to review that. I de-dented background.js because I was thinking about exposing the functions through nodejs so that they could be called from the Python script but I didn't end up doing that yet. Do you remember why you wrapped it in a function? Chrome extensions don't pollute the namespace as far as I can tell.

I think you can merge the redirects, but there's almost certainly a few mistakes and surprises in there.

There's also the issue of that certain redirects might not exist on older Python 3 minor versions, I haven't checked.

verhovsky commented:
I went through all the cases again and organized them better and fixed a few. I also added the rest of the missing 2.6 links and added a simple web page for browsing the special cases more efficiently. Just run `python3 extract_missing_links.py` to download a copy of the documentation locally then open `view-special-cases.html` in a browser and use j/k to navigate.

verhovsky commented:
@m4tx I double checked all the links. Can you review/merge this please?

verhovsky commented:
@m4tx today I wanted to link someone to the part of the Python docs that specify valid python variable names. I googled ["python identifiers"](https://www.google.com/search?q=python+identifiers) and the fifth result (the first result that is the Python docs and not a blog post) was https://docs.python.org/2.5/ref/identifiers.html

with [these changes](https://github.com/m4tx/py3redirect/blob/577dd3112b5b1d4880262e7a18b5e871e8672ad3/special-cases.js#L4117), I would've been correctly redirected to https://docs.python.org/3/reference/lexical_analysis.html#identifiers

Edit: another example, googling "site:docs.python.org function syntax" returns https://docs.python.org/2.0/ref/function.html as the first result, which is currently not redirected, but with [this special case](https://github.com/m4tx/py3redirect/blob/577dd3112b5b1d4880262e7a18b5e871e8672ad3/special-cases.js#L4199) correctly redirects to https://docs.python.org/3/reference/compound_stmts.html#function-definitions

m4tx commented on extract_missing_links.py:
```diff
@@ -0,0 +1,386 @@
+        unzipped_contents = list(temp_path.iterdir())
+        if len(unzipped_contents) == 1:
+            temp_path = unzipped_contents[0]
+        temp_path.rename(output_dir)
```
When not running the script inside the same filesystem as temp location, this line produces the following exception:
```
OSError: [Errno 18] Invalid cross-device link: '/tmp/tmpcnr8gfgx/python_docs' -> 'docs.python.org/2.0'
```
```suggestion
        shutil.move(temp_path, output_dir)
```

m4tx commented on extract_missing_links.py:
```diff
@@ -0,0 +1,386 @@
+import itertools
+import json
+import re
+import string
```
```suggestion
import shutil
import string
```

m4tx commented on extract_missing_links.py:
```diff
@@ -0,0 +1,386 @@
+from pathlib import Path
+import sys
+
+NEW_VERSION = "3.8"
```
```suggestion
NEW_VERSION = "3.9"
```

m4tx commented on extract_missing_links.py:
```diff
@@ -0,0 +1,386 @@
+    "2.5": "https://www.python.org/ftp/python/doc/2.5.4/html-2.5.4.zip",
+    "2.6": "http://docs.python.org/ftp/python/doc/2.6.9/python-2.6.9-docs-html.zip",
+    "2.7": "https://docs.python.org/2/archives/python-2.7.18-docs-html.zip",
+    "3.8": "https://docs.python.org/3/archives/python-3.8.3-docs-html.zip",
```
```suggestion
    "3.9": "https://docs.python.org/3/archives/python-3.9.4-docs-html.zip",
```

m4tx reviewed (commented):
First of all, I am so sorry for not responding for so long. The changes look very nice and I believe the PR is in a state that could be merged pretty much right now. I've added a few comments to the `.py` script; please feel free either to update the special cases with Python 3.9 changes, or just let me know if you want the PR to be merged as is now.

verhovsky commented:
No problem, thanks for finally looking at it! Since I opened this PR, the Python website now redirects moved files to their new place and deleted files to `index.html` https://github.com/python/psf-salt/pull/201 ~~and I think the extension's code might need to be updated to account for that~~. I'll update it to Python 3.9, give me a couple days.

m4tx commented:
@verhovsky are you planning to do more changes, or is it ready to merge now?

verhovsky commented:
I'm not. Unless you want to request some changes, you can merge it.

m4tx commented:
The changes has just been published to Firefox and Chrome addons. Many thanks for this huge contribution! :tada:
